Reorganize the appengine/flexible directory.

jinglundong opened the pull request:
Copy appengine/flexible to appengine/flexible_python37_and_earlier. Move appengine/flexible_ubuntu to appengine/flexible_python38_and_later. We can not remove appengine/flexible yet because our user guide still points to it.

## Description

Fixes #<ISSUE-NUMBER>

Note: Before submitting a pull request, please open an issue for discussion if you are not associated with Google.

## Checklist
- [x] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [x] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [ ] **Tests** pass:   `nox -s py-3.9` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup)). Not done. This is a pure move/refactor. We need to reorganize them as soon as possible. 
- [ ] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup)). Not done. For the same reason. 
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] Please **merge** this PR for me once it is approved.
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample

jinglundong commented:
`Snippet Bot / snippet-bot check` failed, but I don't think I will be able to fix these pre-existing issues in this PR.

dandhlee commented on appengine/flexible_python37_and_earlier/README.md:
```diff
@@ -0,0 +1,73 @@
+
+These are samples for using Python on Google App Engine Flexible Environment. These samples are typically referenced from the [docs](https://cloud.google.com/appengine/docs).
+
+See our other [Google Cloud Platform github repos](https://github.com/GoogleCloudPlatform) for sample applications and
```
nit: any specific repos to take a look?

iennae replied:
@jinglundong I'd suggest just removing this line so it doesn't become a place of technical debt to update regularly.

jinglundong replied:
I don't think we want to point to any specific repos. This is generic enough to be truth for all the code samples. I'd prefer keeping them as is.

dandhlee replied:
We can work on this in a followup, but I'd suggest either

* cut it off early:
```
See our other [Google Cloud Platform github repos](https://github.com/GoogleCloudPlatform) for sample applications
```
or

* give few examples if you're mentioning specific usage like `scaffolding for other frameworks and use cases` since not all samples under `GoogleCloudPlatform` will contain such samples
```
See our other [Google Cloud Platform github repos](https://github.com/GoogleCloudPlatform) for sample applications and scaffolding for other frameworks and use cases, such as {keywords}
```

dandhlee commented on appengine/flexible_python37_and_earlier/README.md:
````diff
@@ -0,0 +1,73 @@
+
+   ```
+   git clone https://github.com/GoogleCloudPlatform/python-docs-samples.git
+   cd python-docs-samples/appengine/flexible
````
```suggestion
   cd python-docs-samples/appengine/flexible_python37_and_earlier
```

jinglundong replied:
Good catch! I will update all of them together.

dandhlee commented on appengine/flexible_python37_and_earlier/README.md:
````diff
@@ -0,0 +1,73 @@
+
+   ```
+   cd hello-world
+   virtualenv env
````
We should use `venv`, see https://cloud.google.com/python/docs/setup

Or alternatively, point to the setup guide.

jinglundong replied:
That makes sense. Please let me know if the updated version looks good to you.

dandhlee commented on appengine/flexible_python37_and_earlier/README.md:
```diff
@@ -0,0 +1,73 @@
+
+4. Congratulations!  Your application is now live at `your-app-id.appspot.com`
+
+## Contributing changes
+
+* See [CONTRIBUTING.md](../CONTRIBUTING.md)
+
+## Licensing
+
+* See [LICENSE](../LICENSE)
```
These links result in a 404.

jinglundong replied:
I think they need to be ../../. Making the change.

dandhlee commented on appengine/flexible_python37_and_earlier/analytics/README.md:
```diff
@@ -0,0 +1,25 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/analytics/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/analytics/README.md
```

jinglundong replied:
Replaced all of them.

dandhlee commented on appengine/flexible_python37_and_earlier/README.md:
```diff
@@ -0,0 +1,73 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/analytics/README.md
```

jinglundong replied:
Done.

dandhlee commented on appengine/flexible_python37_and_earlier/analytics/requirements-test.txt:
```diff
@@ -0,0 +1,3 @@
+pytest==7.0.1
+responses==0.17.0; python_version < '3.7'
```
Fine to remove this now?

jinglundong replied:
I will leave it to the future clean up.

dandhlee commented on appengine/flexible_python37_and_earlier/analytics/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

jinglundong replied:
Future PRs.

dandhlee commented on appengine/flexible_python37_and_earlier/datastore/README.md:
```diff
@@ -0,0 +1,24 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/datastore/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/datastore/README.md
```

jinglundong replied:
Done.

dandhlee commented on appengine/flexible_python37_and_earlier/datastore/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/disk/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/django_cloudsql/README.md:
```diff
@@ -0,0 +1,25 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/django_cloudsql/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/django_cloudsql/README.md
```

dandhlee commented on appengine/flexible_python37_and_earlier/extending_runtime/Dockerfile:
```diff
@@ -0,0 +1,34 @@
+RUN apt-get update && apt-get install -y fortunes
+
+# Change the -p argument to use Python 2.7 if desired.
+RUN virtualenv /env -p python3.4
```
Would it be possible to upgrade this? If not, fine to leave it as is for this PR's scope and update in a followup.

jinglundong replied:
Yes, will leave it to the future PRs.

dandhlee commented on appengine/flexible_python37_and_earlier/extending_runtime/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/hello_world/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/hello_world_django/README.md:
```diff
@@ -0,0 +1,63 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/hello_world_django/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/hello_world_django/README.md
```

dandhlee commented on appengine/flexible_python37_and_earlier/hello_world_django/README.md:
```diff
@@ -0,0 +1,63 @@
+
+This project was created using standard Django commands:
+
+    $ virtualenv env
```
Ditto for `virtualenv`

jinglundong replied:
I added a link to https://cloud.google.com/python/docs/setup. I think it still make sense to keep these historical commands. Their goal was to show how that sample app was created.

dandhlee replied:
Ah. I skimmed through and didn't read properly! Sorry about that. Sounds good to keep it as is.

dandhlee commented on appengine/flexible_python37_and_earlier/hello_world_django/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+Django==4.0.4; python_version >= '3.8'
+Django==3.2.14; python_version < '3.8'
+gunicorn==20.1.0; python_version > '3.0'
+gunicorn==19.10.0; python_version < '3.0'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/metadata/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/multiple_services/README.md:
```diff
@@ -0,0 +1,67 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/multiple_services/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/multiple_services/README.md
```

dandhlee commented on appengine/flexible_python37_and_earlier/multiple_services/README.md:
````diff
@@ -0,0 +1,67 @@
+
+```Bash
+$ cd gateway-service
+$ virtualenv -p python3 env
````
Ditto for `virtualenv`

jinglundong replied:
Done. This is a little bit tricky because we need to ask customers to start two virtual environments.

dandhlee commented on appengine/flexible_python37_and_earlier/multiple_services/gateway-service/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/multiple_services/static-service/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/numpy/requirements.txt:
```diff
@@ -0,0 +1,6 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/pubsub/README.md:
```diff
@@ -0,0 +1,76 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/pubsub/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/pubsub/README.md
```

dandhlee commented on appengine/flexible_python37_and_earlier/pubsub/README.md:
```diff
@@ -0,0 +1,76 @@
+
+Install dependencies, preferably with a virtualenv:
+
+    $ virtualenv env
```
Ditto for `virtualenv`

dandhlee commented on appengine/flexible_python37_and_earlier/pubsub/README.md:
```diff
@@ -0,0 +1,76 @@
+
+2. Create a topic and subscription.
+
+        $ gcloud beta pubsub topics create [your-topic-name]
+        $ gcloud beta pubsub subscriptions create [your-subscription-name] \
+            --topic [your-topic-name] \
+            --push-endpoint \
+                https://[your-app-id].appspot.com/pubsub/push?token=[your-token] \
+            --ack-deadline 30
```
Is it still `gcloud beta pubsub`?

jinglundong replied:
I don't know. Let's keep it for now.

dandhlee commented on appengine/flexible_python37_and_earlier/pubsub/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/scipy/README.md:
```diff
@@ -0,0 +1,9 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/scipy/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/scipy/README.md
```

dandhlee commented on appengine/flexible_python37_and_earlier/scipy/requirements.txt:
```diff
@@ -0,0 +1,9 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/static_files/README.md:
```diff
@@ -0,0 +1,12 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/static_files/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/static_files/README.md
```

dandhlee commented on appengine/flexible_python37_and_earlier/static_files/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/storage/README.md:
```diff
@@ -0,0 +1,37 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/storage/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/storage/README.md
```

dandhlee commented on appengine/flexible_python37_and_earlier/storage/requirements-test.txt:
```diff
@@ -0,0 +1,3 @@
+pytest==7.0.1
+google-cloud-storage==2.0.0; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/storage/requirements.txt:
```diff
@@ -0,0 +1,8 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/tasks/README.md:
```diff
@@ -0,0 +1,115 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/tasks/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/tasks/README.md
```

dandhlee commented on appengine/flexible_python37_and_earlier/tasks/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/twilio/README.md:
```diff
@@ -0,0 +1,34 @@
+[![Open in Cloud Shell][shell_img]][shell_link]
+
+[shell_img]: http://gstatic.com/cloudssh/images/open-btn.png
+[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible/twilio/README.md
```
```suggestion
[shell_link]: https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/GoogleCloudPlatform/python-docs-samples&page=editor&open_in_editor=appengine/flexible_python37_and_earlier/twilio/README.md
```

dandhlee commented on appengine/flexible_python37_and_earlier/twilio/README.md:
```diff
@@ -0,0 +1,34 @@
+
+Before you can run or deploy the sample, you will need to do the following:
+
+1. [Create a Twilio Account](http://ahoy.twilio.com/googlecloudplatform). Google App Engine
```
This link doesn't seem to work anymore (just redirects to twilio.com), but feel free to address in a followup PR instead.

jinglundong replied:
Sorry, this is a pure move of the existing code samples. I have no idea what this link should be. We will need to address that in a future effort.

dandhlee commented on appengine/flexible_python37_and_earlier/twilio/requirements-test.txt:
```diff
@@ -0,0 +1,3 @@
+pytest==7.0.1
+responses==0.17.0; python_version < '3.7'
```
Ditto

dandhlee commented on appengine/flexible_python37_and_earlier/twilio/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+Flask==2.1.0; python_version > '3.6'
+Flask==2.0.3; python_version < '3.7'
```
Ditto

dandhlee reviewed (changes requested):
Reviewing as sample owner mainly for housekeeping items. Please see comments below.

Please feel free to ignore similar comments if they don't apply!

As well, are there any samples we don't want to maintain anymore that can be removed?

dandhlee commented:
Fine to leave the snippet-bot failure as-is. There _might_ be some region tags still using those specific tags.

iennae commented:
hey @jinglundong , it feels like all the setup information should get ripped out where it is multiple places and just point to the setup doc on c.g.c; that way you don't have multiple versions of the setup to update each time.

jinglundong commented:
Thank you so much for all of your comments and suggestions. Please take another look when you have time. I updated all the path to the new one. Updated most of the README files according to the recommendation.

dandhlee reviewed (approved):
LGTM.
